include/xrpl/basics/Number.h

```diff
@@ -13,16 +13,52 @@ class Number;
 std::string
 to_string(Number const& amount);
 
+template <typename T>
+constexpr bool
+isPowerOfTen(T value)
+{
+    while (value >= 10 && value % 10 == 0)
+        value /= 10;
+    return value == 1;
+}
+
 class Number
 {
+public:
+    /** Describes whether and how to enforce this number as an integer.
+     *
+     * - none: No enforcement. The value may vary freely. This is the default.
+     * - compatible: If the absolute value is greater than maxIntValue, valid()
+     *   will return false. Needed for backward compatibility with XRP used in
+     *   AMMs, and available for functions that will do their own checking. This
+     *   is the default for automatic conversions from XRPAmount to Number.
+     * - weak: Like compatible, plus, if the value is unrepresentable (larger
+     *   than maxMantissa), assignment and other operations will throw.
+     * - strong: Like weak, plus, if the absolute value is invalid (larger than
+     *   maxIntValue), assignment and other operations will throw. This is the
+     *   defalut for automatic conversions from MPTAmount to Number.
+     */
+    enum EnforceInteger { none, compatible, weak, strong };
+
+private:
     using rep = std::int64_t;
     rep mantissa_{0};
     int exponent_{std::numeric_limits<int>::lowest()};
 
+    // The enforcement setting is not serialized, and does not affect the
+    // ledger. If not "none", the value is checked to be within the valid
+    // integer range. See the enum description for more detail.
+    EnforceInteger enforceInteger_ = none;
+
 public:
     // The range for the mantissa when normalized
-    constexpr static std::int64_t minMantissa = 1'000'000'000'000'000LL;
-    constexpr static std::int64_t maxMantissa = 9'999'999'999'999'999LL;
+    constexpr static rep minMantissa = 1'000'000'000'000'000LL;
+    static_assert(isPowerOfTen(minMantissa));
+    constexpr static rep maxMantissa = minMantissa * 10 - 1;
+    static_assert(maxMantissa == 9'999'999'999'999'999LL);
+
+    constexpr static rep maxIntValue = maxMantissa / 100;
```

Comment:
how was this number chosen? could we have chosen a larger value?

Reply:
> how was this number chosen? could we have chosen a larger value?

There's a discussion in the comments in https://ripplelabs.atlassian.net/browse/RIPD-4045.

Since that's not public, the tl;dr is that I started with `maxMantissa / 10`, which would give the `STNumber` fields in `Vault` room to grow if they get, for example, extra interest from a special case payment in a `Loan`. I was concerned that wouldn't be enough overhead. The answer was

> Let’s try to go with a safer, conservative limit as in one that reduces the truncation as much as possible

This is intended to be a temporary solution to get the feature available without a major rewrite. A fuller solution should be available in a version or two which allows the entire range of valid numbers.